File: 10.Backtracing/39. Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        combn = []
        totSum = 0
        
        def dfs(i, combn, totSum):
            if totSum == target:
                result.append(combn.copy())
                return
            if i >= len(candidates) or totSum > target:
                return
            
            combn.append(candidates[i])
            # dfs(i+1, combn, totSum + candidates[i])
            dfs(i, combn, totSum + candidates[i])
            
            combn.pop()
            dfs(i+1, combn, totSum)
        
        dfs(0, combn, totSum)
        # dfs(0, combn, 0)
        return result

File: 10.Backtracing/39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        // sort(candidates.begin(), candidates.end()); //若input已sorted，則可有可無
        // dfs(candidates, target, 0, this->totSum);
        dfs(candidates, target, 0, totSum);
        return result;
    }
private:
    vector<vector<int>> result;
    vector<int> combn;
    int totSum=0;
    void dfs(vector<int>& candidates, int target, int i, int totSum) {
        if(totSum == target) {
            result.push_back(combn);
            return;
        }
        if(i >= candidates.size() || totSum > target) {
            return;
        }
        
        combn.push_back(candidates[i]);
        dfs(candidates, target, i, totSum+candidates[i]);
        
        combn.pop_back();
        dfs(candidates, target, i+1, totSum);
        return;
    }
};